### ISSUE TITLE
Legend colorbar is truncated and limits not respected with continuous scales

### ISSUE THREAD

**Issue by Henrik-P** [tidyverse/ggplot2#2650]
Sometimes the legend is not showing the entire range of colours when using a continuous `fill` or `color` scale, e.g `scale_fill_continuous`, `scale_fill_gradient`, `scale_fill_gradientn` (and corresponding `color` scales).

Specifically the upper range of the colorbar legend is truncated, i.e. it doesn't extend all the way to the upper limit of the colour palette.

Furthermore, the title of the legend is vertically adjusted to the (truncated) colorbar, and not to the upper limit.

-----

Set up data, base plot, limits, labels and breaks, adjust `theme` for legend just to make the issue more visible:

``` r
df <- data.frame(x = 1:4, y = 1, col = c(0.1, 0.3, 0.5, 0.7))

library(ggplot2)
p <- ggplot(data = df, aes(x, y, fill = col)) +
  geom_point(size = 15, shape = 21) +
  theme_minimal() +
  theme(legend.background = element_rect(fill = "grey60"),
        legend.text = element_text(size = 18),
        legend.title = element_text(size = 18),
        legend.key.size = unit(1.5, "cm"))

li <- c(0.1, 0.7)
la <- seq(0.1, 0.7, 0.2)
br <- seq(0.1, 0.7, 0.2)

p + scale_fill_continuous(name = "Title", limits = li, labels = la, breaks = br)
```

![](https://i.imgur.com/wKdjyN1.png)

Similar for `scale_fill_gradientn`:

``` r


p + scale_fill_gradientn(colours = c("black", "white"),
                         name = "Title", limits = li, labels = la, breaks = br)
# and scale_fill_gradient
# p + scale_fill_gradient(low = "black", high = "white",
#                         name = "Title", limits = li, labels = la, breaks = br)

```

![](https://i.imgur.com/y2Ly72Y.png)

As you can see, although the largest colour value is 0.7 and the point is correctly coloured on the plot with the end of the colour palette, the colorbar is truncated in the upper range - despite the limits are explicitly set - and the title is in the wrong position.

If you vary the upper limit of the colour scale sometimes it works and sometimes it doesn't.

--------

The following code generates the corresponding issue for `color` scales:

``` r
p <- ggplot(data = df, aes(x, y, color = col)) +
  geom_point(size = 15) +
  theme_minimal() +
  theme(legend.background = element_rect(fill = "grey60"),
        legend.text = element_text(size = 18),
        legend.key.size = unit(1.5, "cm"))

p + scale_color_continuous(limits = li, labels = la, breaks = br)
```

![](https://i.imgur.com/sGdrv99.png)

``` r

p + scale_color_gradientn(colours = c("black", "white"),
                          limits = li, labels = la, breaks = br)
```

![](https://i.imgur.com/W6sIcDp.png)

``` r

p + scale_color_gradient(low = "black", high = "white",
                         limits = li, labels = la, breaks = br)
```

![](https://i.imgur.com/MTIiSE3.png)

-------

[Original post on SO](https://stackoverflow.com/questions/44439423/legend-colorbar-is-truncated-and-limits-not-respected-with-continuous-scales)

**Comment by clauswilke** [comment]
All these examples seem to work with the current codebase. @Henrik-P, could you check whether you can produce the problem at all with the current development version?

``` r
df <- data.frame(x = 1:4, y = 1, col = c(0.1, 0.3, 0.5, 0.7))

library(ggplot2)
p <- ggplot(data = df, aes(x, y, fill = col)) +
  geom_point(size = 15, shape = 21) +
  theme_minimal() +
  theme(legend.background = element_rect(fill = "grey60"),
        legend.text = element_text(size = 18),
        legend.title = element_text(size = 18),
        legend.key.size = unit(1.5, "cm"))

li <- c(0.1, 0.7)
la <- seq(0.1, 0.7, 0.2)
br <- seq(0.1, 0.7, 0.2)

p + scale_fill_continuous(name = "Title", limits = li, labels = la, breaks = br)
```

![](https://i.imgur.com/py0U8fc.png)

``` r
p + scale_fill_gradientn(colours = c("black", "white"),
                         name = "Title", limits = li, labels = la, breaks = br)
```

![](https://i.imgur.com/CLL7lP0.png)

``` r

p <- ggplot(data = df, aes(x, y, color = col)) +
  geom_point(size = 15) +
  theme_minimal() +
  theme(legend.background = element_rect(fill = "grey60"),
        legend.text = element_text(size = 18),
        legend.key.size = unit(1.5, "cm"))

p + scale_color_continuous(limits = li, labels = la, breaks = br)
```

![](https://i.imgur.com/zFE36q8.png)

``` r
p + scale_color_gradientn(colours = c("black", "white"),
                          limits = li, labels = la, breaks = br)
```

![](https://i.imgur.com/2vjBCl4.png)

``` r
p + scale_color_gradient(low = "black", high = "white",
                         limits = li, labels = la, breaks = br)
```

![](https://i.imgur.com/svmq1wP.png)

Created on 2018-05-22 by the [reprex package](http://reprex.tidyverse.org) (v0.2.0).

**Comment by has2k1** [comment]
Fixed in #2346.
